File: bridge/src/ytmd.js
// ytmd.js — client for the ytmdesktop Companion Server (SPEC §2).
//
// Realtime via Socket.IO (NOT raw WS): connect to ws://127.0.0.1:9863/api/v1/realtime,
// websocket-only transport, token in the auth object. The `state-update` event
// carries the full /state object. REST /command is used for control actions; we
// never poll /state for realtime (rate-limited, SPEC §8) — the socket is the feed.
import { io } from "socket.io-client";
import { ensureToken, requestToken } from "./auth.js";
import { YTMD_BASE, YTMD_REALTIME } from "./config.js";

export class YtmdClient {
  constructor({ onState, onConnected, onDisconnected } = {}) {
    this.onState = onState ?? (() => {});
    this.onConnected = onConnected ?? (() => {});
    this.onDisconnected = onDisconnected ?? (() => {});
    this.token = null;
    this.socket = null;
  }

  async start() {
    this.token = await ensureToken();
    this.#connect();
  }

  #connect() {
    this.socket = io(YTMD_REALTIME, {
      transports: ["websocket"], // engine.io polling won't do here (SPEC §2)
      auth: { token: this.token },
      reconnection: true,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 5000,
    });

    this.socket.on("connect", () => {
      console.log("[ytmd] socket connected");
      this.onConnected();
      // Seed current state immediately: the socket only pushes state-update on
      // a CHANGE, so a board that connects (or reconnects) while playback has
      // been sitting idle/paused would otherwise see nothing until the next
      // actual change. One-time fetch on connect, not a recurring poll — the
      // socket remains the feed for everything after this (SPEC §4.2/§8).
      this.#fetchState();
    });

    this.socket.on("state-update", (state) => this.onState(state));

    this.socket.on("disconnect", (reason) => {
      console.log(`[ytmd] socket disconnected: ${reason}`);
      this.onDisconnected();
    });

    this.socket.on("connect_error", async (err) => {
      console.error(`[ytmd] connect_error: ${err.message}`);
      this.onDisconnected();
      // Auth rejected (revoked/overwritten token) -> re-run the handshake once,
      // swap the token, let socket.io retry with it.
      if (/auth|token|unauthor/i.test(err.message)) {
        try {
          this.token = await requestToken();
          this.socket.auth = { token: this.token };
        } catch (e) {
          console.error(`[ytmd] re-auth failed: ${e.message}`);
        }
      }
    });
  }

  // GET /state once per connect, so a paused/idle player still shows correctly
  // for a board that just (re)connected — see the "connect" handler above.
  // Failures are non-fatal: the socket's next real state-update still lands.
  async #fetchState() {
    try {
      const res = await fetch(`${YTMD_BASE}/state`, {
        headers: { Authorization: this.token }, // raw token, no "Bearer" (SPEC §2)
      });
      if (!res.ok) {
        console.warn(`[ytmd] initial /state fetch -> ${res.status}`);
        return;
      }
      this.onState(await res.json());
    } catch (err) {
      console.error(`[ytmd] initial /state fetch failed: ${err.message}`);
    }
  }

  // POST /command {command, data?} with auth header (SPEC §7).
  async command(command, data) {
    const body = data === undefined ? { command } : { command, data };
    let res = await this.#postCommand(body);
    if (res.status === 401) {
      // Token died mid-session — re-auth and retry once.
      this.token = await requestToken();
      if (this.socket) this.socket.auth = { token: this.token };
      res = await this.#postCommand(body);
    }
    if (!res.ok) {
      const text = await res.text().catch(() => "");
      throw new Error(`command ${command} -> ${res.status} ${text}`);
    }
  }

  #postCommand(body) {
    return fetch(`${YTMD_BASE}/command`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: this.token, // raw token, no "Bearer" (SPEC §2)
      },
      body: JSON.stringify(body),
    });
  }

  stop() {
    this.socket?.close();
  }
}

File: app/bridge-core/src/ytmd.rs
//! Client for the ytmdesktop Companion Server (ports ytmd.js).
//! Realtime via Socket.IO (websocket-only), token in the auth object; the
//! `state-update` event carries the full /state object. REST /state seeds once
//! on connect (socket only pushes on change); /command drives control actions.
use crate::commands::YtmdCommand;
use crate::config::{ytmd_base, ytmd_realtime_base, ytmd_realtime_namespace};
use anyhow::{Context, Result};
use futures_util::FutureExt;
use rust_socketio::asynchronous::{Client, ClientBuilder};
use rust_socketio::{Event, Payload, TransportType};
use serde_json::{json, Value};
use tokio::sync::mpsc::UnboundedSender;

/// True when a socket/command error message indicates the token was rejected
/// (revoked/expired/overwritten). Ports the JS `/auth|token|unauthor/i` test.
fn is_auth_error(msg: &str) -> bool {
    let m = msg.to_ascii_lowercase();
    ["auth", "token", "unauthor"].iter().any(|kw| m.contains(kw))
}

pub struct YtmdHandle {
    socket: Client,
    http: reqwest::Client,
    token: String,
}

pub async fn connect(
    http: reqwest::Client,
    token: String,
    on_state: UnboundedSender<Value>,
    on_connect: UnboundedSender<()>,
    on_disconnect: UnboundedSender<()>,
) -> Result<YtmdHandle> {
    let state_tx = on_state.clone();
    let state_cb = move |payload: Payload, _: Client| {
        let tx = state_tx.clone();
        async move {
            if let Payload::Text(values) = payload {
                if let Some(state) = values.into_iter().next() {
                    let _ = tx.send(state);
                }
            }
        }
        .boxed()
    };

    let conn_tx = on_connect.clone();
    let connect_cb = move |_: Payload, _: Client| {
        let tx = conn_tx.clone();
        async move {
            let _ = tx.send(());
        }
        .boxed()
    };

    let dis_tx = on_disconnect.clone();
    let close_cb = move |_: Payload, _: Client| {
        let tx = dis_tx.clone();
        async move {
            let _ = tx.send(());
        }
        .boxed()
    };

    let socket = ClientBuilder::new(ytmd_realtime_base())
        .namespace(ytmd_realtime_namespace())
        .auth(json!({ "token": token }))
        .transport_type(TransportType::Websocket)
        // Mirror the JS client (bridge/src/ytmd.js): reconnection:true,
        // reconnectionDelay:1000, reconnectionDelayMax:5000. reconnect_on_disconnect
        // ensures a SERVER-initiated disconnect also triggers reconnection.
        .reconnect(true)
        .reconnect_on_disconnect(true)
        .reconnect_delay(1000, 5000)
        .on("state-update", state_cb)
        .on(Event::Connect, connect_cb)
        .on(Event::Close, close_cb)
        .on(Event::Error, |err, _: Client| {
            async move { tracing::error!("[ytmd] socket error: {err:?}") }.boxed()
        })
        .connect()
        .await
        .context("Socket.IO connect failed")?;

    Ok(YtmdHandle { socket, http, token })
}

impl YtmdHandle {
    /// One-shot GET /state so a paused/idle player still paints for a board that
    /// just (re)connected. Failures are non-fatal.
    pub async fn fetch_state_once(&self) -> Option<Value> {
        let url = format!("{}/state", ytmd_base());
        match self
            .http
            .get(&url)
            .header("Authorization", &self.token) // raw token, no "Bearer"
            .send()
            .await
            .and_then(|r| r.error_for_status())
        {
            Ok(r) => r.json().await.ok(),
            Err(e) => {
                tracing::warn!("[ytmd] initial /state fetch failed: {e}");
                None
            }
        }
    }

    pub async fn send_command(&self, c: &YtmdCommand) -> Result<()> {
        let url = format!("{}/command", ytmd_base());
        let body = match &c.data {
            Some(d) => json!({ "command": c.command, "data": d }),
            None => json!({ "command": c.command }),
        };
        self.http
            .post(&url)
            .header("Authorization", &self.token)
            .json(&body)
            .send()
            .await?
            .error_for_status()
            .with_context(|| format!("command {}", c.command))?;
        Ok(())
    }

    pub async fn disconnect(self) {
        let _ = self.socket.disconnect().await;
    }
}

#[cfg(test)]
mod tests {
    use super::is_auth_error;

    #[test]
    fn is_auth_error_matches_auth_keywords() {
        assert!(is_auth_error("auth failed"));
        assert!(is_auth_error("invalid token"));
        assert!(is_auth_error("Unauthorized"));
        assert!(is_auth_error("UNAUTHORISED")); // case-insensitive
        assert!(is_auth_error("Received an ConnectError frame: token rejected"));
    }

    #[test]
    fn is_auth_error_ignores_transient_errors() {
        assert!(!is_auth_error("transport close"));
        assert!(!is_auth_error("timeout"));
        assert!(!is_auth_error("connection refused"));
        assert!(!is_auth_error(""));
    }
}
